### PR TITLE
Implement Non-Strict Affinity and  Non-Strict Anti-Affinity Support

### PR DESCRIPTION
**Pull request by FarnazBGH**
**_*Description of changes:*_**
This PR integrates non-strict Anti-affinity and non-strict Affinity types into CloudStackAffinityGroup. With the latest CloudStack version 4.18, we now have the capability to set these flexible affinity rules. This enhancement allows for more adaptable and efficient resource management, particularly beneficial for the configuration of worker nodes.

**_*Testing performed:*_**

* Make Test
* Creating cluster with similar `CloudStackMachineTemplate` for the worker machines. 
Below is an example configuration showcasing the use of a non-strict anti-affinity type in `CloudStackMachineTemplate`:
  ```yaml
  apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
  kind: CloudStackMachineTemplate
  metadata:
    name: example-worker-0
  spec:
    template:
      spec:
        affinity: soft-anti
        offering:
          name: example-offering
        template:
          name: example-template
  ```

_**Requirements:**_
**CloudStack Version**: This feature requires CloudStack version 4.18.1 or higher.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by g-gaston — COMMENTED** [review]
Thanks a lot for the PR @FarnazBGH!

How would this behave if the user is running an older version of CloudStack that doesn't support this feature?

**Comment by FarnazBGH** [comment]
> Thanks a lot for the PR @FarnazBGH!
> 
> How would this behave if the user is running an older version of CloudStack that doesn't support this feature?

In our environment all the CloudStack are upgraded to the latest version so non-strict affinities are available, for respond to your question I created an extra-test affinity (assume there is affinity in the later version which is test-anti) and revised the code with test affinity which is not available in our system which was (TestAntiAffinity = "test-anti"):
```
const (
	// The presence of a finalizer prevents CAPI from deleting the corresponding CAPI data.
	MachineFinalizer = "cloudstackmachine.infrastructure.cluster.x-k8s.io"
	ProAffinity      = "pro"
	AntiAffinity     = "anti"
	SoftProAffinity  = "soft-pro"
	SoftAntiAffinity = "soft-anti"
	TestAntiAffinity = "test-anti"
	NoAffinity       = "no"
)
```

Now during the creation of the cluster when I set the affinity to the value of"test-anti", I will get the below error:
```
E0109 09:53:14.109649       1 controller.go:324]  "msg"="Reconciler error" "error"="CloudStack API error 431 (CSExceptionErrorCode: 4350): Unable to create affinity group, invalid affinity group type: test host anti-affinity. Valid values are non-strict host affinity,host affinity,host anti-affinity,ExplicitDedication,non-strict host anti-affinity" 
```

which means for the lower version of cloudstack they will get similar errors for the non-strict affinities.

**Comment by davidjumani** [comment]
LGTM but waiting for input from other maintainers as well.
@g-gaston If this is tied to a specific version of cloudstack, the [compatibility matrix](https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack#compatibility-with-apache-cloudstack-versions) can be updated accordingly when a release with this feature is cut so users are aware of this requirement

**Review by hrak — CHANGES_REQUESTED** [review]
Left some small comments. 

Question from my side to the other maintainers would be whether its a good idea to apply this change to v1beta1 and v1beta2, or if it doesn't really matter because the type of value is not changing, just the allowed values.

And a general suggestion for this Type field from my side would be to consider using the following kubebuilder validation so the CRD already defines the allowed values:

In `cloudstackaffinitygroup_types.go`:
```golang
type CloudStackAffinityGroupSpec struct {
	// Mutually exclusive parameter with AffinityGroupIDs.
	// Can be "host affinity" or "host anti-affinity". Will create an affinity group per machine set.
	// +kubebuilder:validation:Enum=host affinity;host anti-affinity;non-strict host affinity;non-strict host anti-affinity
	// +optional
	Type string `json:"type,omitempty"`
```

In `cloudstackmachine_types.go`:
```golang
	// Mutually exclusive parameter with AffinityGroupIDs.
	// Defaults to `no`. Can be `pro` or `anti`. Will create an affinity group per machine set.
	// +kubebuilder:validation:Enum=no;pro;anti;soft-pro;soft-anti
	// +optional
	Affinity string `json:"affinity,omitempty"`
```

Another question from my end is why we use two different formats for the affinity types (`host affinity` vs `pro` f.e.), but this might be something we can deal with in a seperate PR.

**Comment by g-gaston** [comment]
> LGTM but waiting for input from other maintainers as well. @g-gaston If this is tied to a specific version of cloudstack, the [compatibility matrix](https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack#compatibility-with-apache-cloudstack-versions) can be updated accordingly when a release with this feature is cut so users are aware of this requirement

Totally. However, I would prefer not to drop support for older versions (for now) just to support this small feature.

I think this brings us to the "partial support" territory: not all the features work in all supported versions. I think as long as we document clearly what we support and what we not and what is the expected behavior (if you use this feature in older versions, validations will pass but you'll get a reconciler error), this is a reasonable approach.

Opinions?

@FarnazBGH could you add a line to the go comment saying that these new values are only supported in CS 4.18.1 or higher?
And we need to find a place to document this in the book/docs. Do you mind taking a look and proposing a change for it?

File: test/e2e/affinity_group.go [hunk 1]
```diff
@@ -73,6 +73,14 @@ func AffinityGroupSpec(ctx context.Context, inputGetter func() CommonSpecInput)
...
 	})
 
+	It("Should have host affinity group when affinity is soft-pro", func() {
+		affinityIds = executeTest(ctx, input, namespace, specName, clusterResources, "soft-pro")
+	})
+
+	It("Should have host affinity group when affinity is soft-anti", func() {
+		affinityIds = executeTest(ctx, input, namespace, specName, clusterResources, "soft-anti")
+	})
+
```
Comment (vishesh92) [inline test/e2e/affinity_group.go:83]: @FarnazBGH Can you add a check for running this only for supported cloudstack versions else skip?

**Comment by vishesh92** [comment]
> I think this brings us to the "partial support" territory: not all the features work in all supported versions. I think as long as we document clearly what we support and what we not and what is the expected behavior (if you use this feature in older versions, validations will pass but you'll get a reconciler error), this is a reasonable approach.
> 
> Opinions?
> 

Agree with @g-gaston . If the affinity group is not supported, reconciler error should be fine.

**Comment by vishesh92** [comment]
> @FarnazBGH could you add a line to the go comment saying that these new values are only supported in CS 4.18.1 or higher? And we need to find a place to document this in the book/docs. Do you mind taking a look and proposing a change for it?

@FarnazBGH Can you update the documentation about the new affinity group types and add a note about ACS version compatability in https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack/blob/main/docs/book/src/clustercloudstack/configuration.md#affinity-groups?

**Comment by vishesh92** [comment]
@FarnazBGH I am planning for a release next week. Do you think this PR will be ready by then?

**Comment by FarnazBGH** [comment]
Sorry for the delayed response.

The comments and documentation are added.

 In the next few days, I will also work on checking and testing the CloudStack version for the soft-pro and soft-anti.

**Comment by FarnazBGH** [comment]
The cloudStack version Check for soft affinities is added. The result for cloudStack Versions lower than 4.18 will be:

```
S [SKIPPING] [0.492 seconds]
When testing affinity group
~/cluster-api-provider-cloudstack/test/e2e/affinity_group_test.go:27
  Should have host affinity group when affinity is soft-pro [It]
~/cluster-api-provider-cloudstack/test/e2e/affinity_group.go:76

  Soft affinity groups are only supported in CloudStack version 4.18 or higher.

~/cluster-api-provider-cloudstack/test/e2e/affinity_group.go:79
------------------------------
When testing affinity group
  Should have host affinity group when affinity is soft-anti
~/cluster-api-provider-cloudstack/test/e2e/affinity_group.go:84
STEP: Getting a CloudStack client secret
STEP: Creating a CloudStack client
STEP: Creating a namespace for hosting the "affinity-group" test spec
INFO: Creating namespace affinity-group-psmshm
INFO: Creating event watcher for namespace "affinity-group-psmshm"

S [SKIPPING] [0.417 seconds]
When testing affinity group
~/cluster-api-provider-cloudstack/test/e2e/affinity_group_test.go:27
  Should have host affinity group when affinity is soft-anti [It]
~//cluster-api-provider-cloudstack/test/e2e/affinity_group.go:84

  Soft affinity groups are only supported in CloudStack version 4.18 or higher.

~/cluster-api-provider-cloudstack/test/e2e/affinity_group.go:87
```

File: test/e2e/affinity_group.go (hunk 1 above)
Comment (FarnazBGH) [reply test/e2e/affinity_group.go]: The cloudStack version Check for soft affinities is added. The result for cloudStack Versions lower than 4.18 will be:

```
S [SKIPPING] [0.492 seconds]
When testing affinity group
~/cluster-api-provider-cloudstack/test/e2e/affinity_group_test.go:27
  Should have host affinity group when affinity is soft-pro [It]
~/cluster-api-provider-cloudstack/test/e2e/affinity_group.go:76

  Soft affinity groups are only supported in CloudStack version 4.18 or higher.

~/cluster-api-provider-cloudstack/test/e2e/affinity_group.go:79
------------------------------
When testing affinity group
  Should have host affinity group when affinity is soft-anti
~/cluster-api-provider-cloudstack/test/e2e/affinity_group.go:84
STEP: Getting a CloudStack client secret
STEP: Creating a CloudStack client
STEP: Creating a namespace for hosting the "affinity-group" test spec
INFO: Creating namespace affinity-group-psmshm
INFO: Creating event watcher for namespace "affinity-group-psmshm"

S [SKIPPING] [0.417 seconds]
When testing affinity group
~/cluster-api-provider-cloudstack/test/e2e/affinity_group_test.go:27
  Should have host affinity group when affinity is soft-anti [It]
~//cluster-api-provider-cloudstack/test/e2e/affinity_group.go:84

  Soft affinity groups are only supported in CloudStack version 4.18 or higher.

~/cluster-api-provider-cloudstack/test/e2e/affinity_group.go:87
```

File: test/e2e/affinity_group.go [hunk 2]
```diff
@@ -73,6 +73,22 @@ func AffinityGroupSpec(ctx context.Context, inputGetter func() CommonSpecInput)
 		affinityIds = executeTest(ctx, input, namespace, specName, clusterResources, "anti")
 	})
 
+	It("Should have host affinity group when affinity is soft-pro", func() {
+		cloudStackVersion := input.E2EConfig.GetVariable("CLOUDSTACK_VERSION")
```
Comment (vishesh92) [inline test/e2e/affinity_group.go:77]: Can we make this dynamic and fetch the version from CloudStack itself?
The below code is not merged, but maybe something like this?

https://github.com/shapeblue/cluster-api-provider-cloudstack/blob/e96ede3956bce5df9fdb9c3306c498dd545ec07b/test/e2e/kubernetes_cks_cluster.go#L51-L68

https://github.com/shapeblue/cluster-api-provider-cloudstack/blob/e96ede3956bce5df9fdb9c3306c498dd545ec07b/test/e2e/common.go#L254-L263

File: pkg/cloud/affinity_groups.go [hunk 3]
```diff
@@ -22,8 +22,10 @@ import (
 )
 
 const (
-	AntiAffinityGroupType = "host anti-affinity"
-	AffinityGroupType     = "host affinity"
+	AntiAffinityGroupType     = "host anti-affinity"
+	AffinityGroupType         = "host affinity"
+	SoftAntiAffinityGroupType = "non-strict anti-affinity"
```
Comment (vishesh92) [inline pkg/cloud/affinity_groups.go:27]:
```suggestion
	SoftAntiAffinityGroupType = "non-strict host anti-affinity"
```

Comment (vishesh92) [reply pkg/cloud/affinity_groups.go:27]: @FarnazBGH This needs to be fixed.

**Comment by vishesh92** [comment]
@FarnazBGH I will be cutting the RC on Monday. If possible, please address/resolve the comments before then. Else we will have to move this to the next release.

**Pull request closed without merging**